File: src/MicrosoftAuth.cs
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace PetusLauncher;

// Microsoft account auth for Minecraft. Full device-code → Xbox Live → XSTS →
// Minecraft services chain, plus silent refresh and offline accounts.
// Uses the public MC client id (Config.MsClientId), device-code flow, so no
// client secret or redirect server is needed.
static class MicrosoftAuth
{
    const string DeviceCodeUrl = "https://login.microsoftonline.com/consumers/oauth2/v2.0/devicecode";
    const string TokenUrl = "https://login.microsoftonline.com/consumers/oauth2/v2.0/token";
    const string XblUrl = "https://user.auth.xboxlive.com/user/authenticate";
    const string XstsUrl = "https://xsts.auth.xboxlive.com/xsts/authorize";
    const string McLoginUrl = "https://api.minecraftservices.com/authentication/login_with_xbox";
    const string McProfileUrl = "https://api.minecraftservices.com/minecraft/profile";
    const string Scope = "XboxLive.signin offline_access";

    static readonly HttpClient Http = MakeClient();

    static HttpClient MakeClient()
    {
        var c = new HttpClient { Timeout = TimeSpan.FromSeconds(60) };
        c.DefaultRequestHeaders.UserAgent.ParseAdd("PetusLauncher/2.1");
        c.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        return c;
    }

    // Device-code login. showCode(userCode, verificationUri) lets the UI tell the
    // user where to go. Polls until authorized, runs the full chain, returns a
    // ready Microsoft McAccount. Throws on failure / timeout / expiry.
    public static async Task<McAccount> LoginAsync(Action<string, string> showCode, CancellationToken ct)
    {
        // 1) Request a device code.
        using var codeResp = await Http.PostAsync(DeviceCodeUrl, Form(new()
        {
            ["client_id"] = Config.MsClientId,
            ["scope"] = Scope,
        }), ct);
        var codeBody = await codeResp.Content.ReadAsStringAsync(ct);
        if (!codeResp.IsSuccessStatusCode)
        {
            // Surface Microsoft's real reason instead of a generic 400 so the
            // error is actionable (usually: the client_id isn't authorized for
            // the device-code flow / consumers endpoint).
            string detail = codeBody;
            try
            {
                var e = JsonDocument.Parse(codeBody).RootElement;
                detail = (e.TryGetProperty("error_description", out var d) ? d.GetString() : null)
                       ?? (e.TryGetProperty("error", out var er) ? er.GetString() : null)
                       ?? codeBody;
            }
            catch { }
            throw new Exception($"Microsoft отклонил запрос ({(int)codeResp.StatusCode}): {detail}");
        }
        var root = JsonDocument.Parse(codeBody).RootElement;
        var deviceCode = root.GetProperty("device_code").GetString()!;
        var userCode = root.GetProperty("user_code").GetString()!;
        var verifyUri = root.GetProperty("verification_uri").GetString()!;
        int interval = root.TryGetProperty("interval", out var iv) ? iv.GetInt32() : 5;
        int expiresIn = root.TryGetProperty("expires_in", out var ex) ? ex.GetInt32() : 900;

        showCode(userCode, verifyUri);

        // 2) Poll the token endpoint until the user authorizes (or it expires).
        var deadline = DateTime.UtcNow.AddSeconds(expiresIn);
        (string accessToken, string refreshToken) ms;
        while (true)
        {
            ct.ThrowIfCancellationRequested();
            if (DateTime.UtcNow >= deadline)
                throw new TimeoutException("Время ожидания входа истекло. Попробуйте снова.");
            await Task.Delay(TimeSpan.FromSeconds(interval), ct);

            using var poll = await Http.PostAsync(TokenUrl, Form(new()
            {
                ["client_id"] = Config.MsClientId,
                ["grant_type"] = "urn:ietf:params:oauth:grant-type:device_code",
                ["device_code"] = deviceCode,
            }), ct);
            using var pollJson = await ParseJson(poll);
            var pr = pollJson.RootElement;

            if (poll.IsSuccessStatusCode)
            {
                ms = (pr.GetProperty("access_token").GetString()!,
                      pr.GetProperty("refresh_token").GetString()!);
                break;
            }

            var error = pr.TryGetProperty("error", out var e) ? e.GetString() : "unknown";
            if (error == "authorization_pending") continue;      // user hasn't finished yet
            if (error == "slow_down") { interval += 5; continue; } // back off as requested
            if (error == "authorization_declined") throw new Exception("Вход отклонён пользователем.");
            if (error == "expired_token") throw new TimeoutException("Код устарел. Попробуйте снова.");
            throw new Exception($"Ошибка входа Microsoft: {error}");
        }

        return await FinishChainAsync(ms.accessToken, ms.refreshToken, ct);
    }

    // Silent refresh from a stored refresh token. Returns null when the token is
    // no longer valid (user must log in again). Non-Microsoft accounts → null.
    public static async Task<McAccount?> RefreshAsync(McAccount acc)
    {
        if (acc.Type != "microsoft" || string.IsNullOrEmpty(acc.RefreshToken)) return null;
        try
        {
            using var resp = await Http.PostAsync(TokenUrl, Form(new()
            {
                ["client_id"] = Config.MsClientId,
                ["grant_type"] = "refresh_token",
                ["refresh_token"] = acc.RefreshToken,
                ["scope"] = Scope,
            }));
            if (!resp.IsSuccessStatusCode) return null;
            using var json = await ParseJson(resp);
            var r = json.RootElement;
            var access = r.GetProperty("access_token").GetString()!;
            var refresh = r.TryGetProperty("refresh_token", out var rt) ? rt.GetString()! : acc.RefreshToken;
            return await FinishChainAsync(access, refresh, CancellationToken.None);
        }
        catch { return null; }
    }

    // Offline account: name only. UUID is a name-based v3 (MD5) UUID over
    // "OfflinePlayer:<name>", matching the vanilla offline-mode convention.
    public static McAccount OfflineAccount(string name)
    {
        var uuid = NameUuidV3("OfflinePlayer:" + name);
        return new McAccount(name, uuid, "0", "offline", 0, "");
    }

    // -------------------------------------------------------- the token chain

    // MS access token → XBL → XSTS → Minecraft → profile → McAccount.
    static async Task<McAccount> FinishChainAsync(string msAccess, string refreshToken, CancellationToken ct)
    {
        // Xbox Live user token.
        var (xblToken, uhs) = await XblAuthAsync(msAccess, ct);
        // XSTS token (authorizes for the Minecraft relying party).
        var xstsToken = await XstsAuthAsync(xblToken, ct);
        // Minecraft services access token.
        var mcToken = await McLoginAsync(uhs, xstsToken, ct);
        // Profile (name + uuid). Fails if the account owns no copy of the game.
        var (name, uuid) = await McProfileAsync(mcToken, ct);

        // MS refresh tokens live ~90 days; store an approximate expiry.
        long refreshExpires = DateTimeOffset.UtcNow.AddDays(80).ToUnixTimeSeconds();
        return new McAccount(name, uuid, mcToken, "microsoft", refreshExpires, refreshToken);
    }

    static async Task<(string token, string uhs)> XblAuthAsync(string msAccess, CancellationToken ct)
    {
        var body = new
        {
            Properties = new
            {
                AuthMethod = "RPS",
                SiteName = "user.auth.xboxlive.com",
                RpsTicket = "d=" + msAccess, // 'd=' prefix required for MSA tokens
            },
            RelyingParty = "http://auth.xboxlive.com",
            TokenType = "JWT",
        };
        using var resp = await Http.PostAsync(XblUrl, JsonContent(body), ct);
        if (!resp.IsSuccessStatusCode)
            throw new Exception("Xbox Live отклонил вход (XBL).");
        using var json = await ParseJson(resp);
        var r = json.RootElement;
        var token = r.GetProperty("Token").GetString()!;
        var uhs = r.GetProperty("DisplayClaims").GetProperty("xui")[0].GetProperty("uhs").GetString()!;
        return (token, uhs);
    }

    static async Task<string> XstsAuthAsync(string xblToken, CancellationToken ct)
    {
        var body = new
        {
            Properties = new
            {
                SandboxId = "RETAIL",
                UserTokens = new[] { xblToken },
            },
            RelyingParty = "rp://api.minecraftservices.com/",
            TokenType = "JWT",
        };
        using var resp = await Http.PostAsync(XstsUrl, JsonContent(body), ct);
        using var json = await ParseJson(resp);
        var r = json.RootElement;
        if (!resp.IsSuccessStatusCode)
        {
            // Known XSTS error codes → friendly messages.
            long xerr = r.TryGetProperty("XErr", out var xe) ? xe.GetInt64() : 0;
            throw xerr switch
            {
                2148916233 => new Exception("У этого аккаунта Microsoft нет профиля Xbox. Создайте его на xbox.com."),
                2148916235 => new Exception("Xbox Live недоступен в вашем регионе."),
                2148916236 or 2148916237 => new Exception("Требуется подтверждение возраста для этого аккаунта."),
                2148916238 => new Exception("Детский аккаунт: добавьте его в семейную группу Microsoft."),
                _ => new Exception($"Ошибка XSTS ({xerr}).")
            };
        }
        return r.GetProperty("Token").GetString()!;
    }

    static async Task<string> McLoginAsync(string uhs, string xstsToken, CancellationToken ct)
    {
        var body = new { identityToken = $"XBL3.0 x={uhs};{xstsToken}" };
        using var resp = await Http.PostAsync(McLoginUrl, JsonContent(body), ct);
        if (!resp.IsSuccessStatusCode)
            throw new Exception("Не удалось войти в сервисы Minecraft.");
        using var json = await ParseJson(resp);
        return json.RootElement.GetProperty("access_token").GetString()!;
    }

    static async Task<(string name, string uuid)> McProfileAsync(string mcToken, CancellationToken ct)
    {
        using var req = new HttpRequestMessage(HttpMethod.Get, McProfileUrl);
        req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", mcToken);
        using var resp = await Http.SendAsync(req, ct);
        if (!resp.IsSuccessStatusCode)
            throw new Exception("На этом аккаунте нет купленной копии Minecraft.");
        using var json = await ParseJson(resp);
        var r = json.RootElement;
        var name = r.GetProperty("name").GetString()!;
        var id = r.GetProperty("id").GetString()!; // 32 hex chars, no dashes
        return (name, id);
    }

    // ----------------------------------------------------------- small helpers

    static FormUrlEncodedContent Form(Dictionary<string, string> d) => new(d);

    static StringContent JsonContent(object o)
        => new(JsonSerializer.Serialize(o), Encoding.UTF8, "application/json");

    static async Task<JsonDocument> ReadJson(HttpResponseMessage resp)
    {
        resp.EnsureSuccessStatusCode();
        return JsonDocument.Parse(await resp.Content.ReadAsStringAsync());
    }

    // Parse a response body as JSON regardless of status (callers inspect errors).
    static async Task<JsonDocument> ParseJson(HttpResponseMessage resp)
        => JsonDocument.Parse(await resp.Content.ReadAsStringAsync());

    // Name-based v3 (MD5) UUID, RFC 4122: set version + variant bits.
    static string NameUuidV3(string input)
    {
        byte[] hash = MD5.HashData(Encoding.UTF8.GetBytes(input));
        hash[6] = (byte)((hash[6] & 0x0F) | 0x30); // version 3
        hash[8] = (byte)((hash[8] & 0x3F) | 0x80); // RFC 4122 variant
        var hex = Convert.ToHexString(hash).ToLowerInvariant();
        return $"{hex[..8]}-{hex[8..12]}-{hex[12..16]}-{hex[16..20]}-{hex[20..]}";
    }
}

File: src/main/java/ru/petus/launcher/auth/MicrosoftAuth.java
package ru.petus.launcher.auth;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import ru.petus.launcher.core.Json;
import ru.petus.launcher.core.Log;
import ru.petus.launcher.core.Settings;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Microsoft (premium) login via the OAuth 2.0 device code flow — the only flow
 * that works well in a desktop app without shipping a client secret or an
 * embedded browser: we show a code, the user confirms it in a browser, we poll.
 *
 * Chain: MSA → Xbox Live → XSTS → Minecraft services → profile.
 *
 * The Azure application id is a setting (Settings.microsoftClientId): every
 * launcher must register its own public client with the XboxLive.signin scope.
 */
public final class MicrosoftAuth {
    public static final String DEFAULT_SCOPE = "XboxLive.signin offline_access";

    private static final String DEVICE_CODE_URL =
            "https://login.microsoftonline.com/consumers/oauth2/v2.0/devicecode";
    private static final String TOKEN_URL =
            "https://login.microsoftonline.com/consumers/oauth2/v2.0/token";
    private static final String XBL_URL = "https://user.auth.xboxlive.com/user/authenticate";
    private static final String XSTS_URL = "https://xsts.auth.xboxlive.com/xsts/authorize";
    private static final String MC_LOGIN_URL =
            "https://api.minecraftservices.com/authentication/login_with_xbox";
    private static final String MC_PROFILE_URL = "https://api.minecraftservices.com/minecraft/profile";
    private static final String MC_ENTITLEMENTS_URL = "https://api.minecraftservices.com/entitlements/mcstore";

    private final HttpClient http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(15))
            .build();

    /** Details shown to the user while the device code flow is pending. */
    public record DeviceCode(String userCode, String verificationUri, String deviceCode, int interval,
            int expiresIn, String message) {
    }

    public static final class AuthException extends RuntimeException {
        public AuthException(String message) {
            super(message);
        }

        public AuthException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    public static boolean configured() {
        return !Settings.get().microsoftClientId.isBlank();
    }

    private String clientId() {
        String clientId = Settings.get().microsoftClientId;
        if (clientId.isBlank()) {
            throw new AuthException("Не задан Azure Client ID для входа через Microsoft. "
                    + "Укажите его в Настройках → Аккаунты.");
        }
        return clientId;
    }

    // --- low level helpers ----------------------------------------------
    private JsonObject postForm(String url, Map<String, String> form) {
        StringBuilder body = new StringBuilder();
        form.forEach((key, value) -> {
            if (body.length() > 0) {
                body.append('&');
            }
            body.append(URLEncoder.encode(key, StandardCharsets.UTF_8)).append('=')
                    .append(URLEncoder.encode(value, StandardCharsets.UTF_8));
        });
        return send(HttpRequest.newBuilder(URI.create(url))
                .header("Content-Type", "application/x-www-form-urlencoded")
                .timeout(Duration.ofSeconds(30))
                .POST(HttpRequest.BodyPublishers.ofString(body.toString(), StandardCharsets.UTF_8))
                .build(), true);
    }

    private JsonObject postJson(String url, JsonObject payload, String bearer) {
        HttpRequest.Builder request = HttpRequest.newBuilder(URI.create(url))
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .timeout(Duration.ofSeconds(30))
                .POST(HttpRequest.BodyPublishers.ofString(payload.toString(), StandardCharsets.UTF_8));
        if (bearer != null) {
            request.header("Authorization", "Bearer " + bearer);
        }
        return send(request.build(), false);
    }

    private JsonObject getJson(String url, String bearer) {
        return send(HttpRequest.newBuilder(URI.create(url))
                .header("Accept", "application/json")
                .header("Authorization", "Bearer " + bearer)
                .timeout(Duration.ofSeconds(30))
                .GET()
                .build(), false);
    }

    private JsonObject send(HttpRequest request, boolean allowErrorBody) {
        try {
            HttpResponse<String> response = http.send(request,
                    HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            JsonObject object = Json.parseObject(response.body());
            if (response.statusCode() / 100 != 2 && !allowErrorBody) {
                throw new AuthException("Microsoft ответил " + response.statusCode() + ": "
                        + Json.string(object, "errorMessage", response.body()));
            }
            return object;
        } catch (IOException error) {
            throw new AuthException("Нет связи с серверами Microsoft: " + error.getMessage(), error);
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new AuthException("Вход прерван");
        }
    }

    // --- flow ------------------------------------------------------------
    public DeviceCode requestDeviceCode() {
        JsonObject response = postForm(DEVICE_CODE_URL, Map.of(
                "client_id", clientId(),
                "scope", DEFAULT_SCOPE));
        String deviceCode = Json.string(response, "device_code", "");
        if (deviceCode.isEmpty()) {
            throw new AuthException("Microsoft не выдал код устройства: "
                    + Json.string(response, "error_description", "неизвестная ошибка"));
        }
        return new DeviceCode(
                Json.string(response, "user_code", ""),
                Json.string(response, "verification_uri", "https://microsoft.com/link"),
                deviceCode,
                Math.max(1, Json.number(response, "interval", 5)),
                Json.number(response, "expires_in", 900),
                Json.string(response, "message", ""));
    }

    /** Polls until the user confirms the code, then completes the whole chain. */
    public Account completeDeviceCode(DeviceCode code, Consumer<String> progress) {
        long deadline = System.currentTimeMillis() + code.expiresIn() * 1000L;
        int interval = code.interval();
        while (System.currentTimeMillis() < deadline) {
            sleep(interval);
            JsonObject response = postForm(TOKEN_URL, Map.of(
                    "client_id", clientId(),
                    "grant_type", "urn:ietf:params:oauth:grant-type:device_code",
                    "device_code", code.deviceCode()));
            String error = Json.string(response, "error", "");
            if (error.isEmpty()) {
                progress.accept("Подтверждено. Получаем сессию Minecraft…");
                return finishLogin(Json.string(response, "access_token", ""),
                        Json.string(response, "refresh_token", ""), progress);
            }
            switch (error) {
                case "authorization_pending" -> progress.accept("Ждём подтверждения в браузере…");
                case "slow_down" -> interval += 5;
                case "authorization_declined" -> throw new AuthException("Вы отклонили вход в Microsoft");
                case "expired_token" -> throw new AuthException("Код истёк. Попробуйте ещё раз.");
                default -> throw new AuthException("Microsoft: "
                        + Json.string(response, "error_description", error));
            }
        }
        throw new AuthException("Время ожидания входа истекло");
    }

    /** Silent re-login for a stored premium account. */
    public Account refresh(Account account) {
        if (account.msaRefreshToken == null || account.msaRefreshToken.isBlank()) {
            throw new AuthException("Нет refresh-токена Microsoft — войдите в аккаунт заново");
        }
        JsonObject response = postForm(TOKEN_URL, Map.of(
                "client_id", clientId(),
                "grant_type", "refresh_token",
                "refresh_token", account.msaRefreshToken,
                "scope", DEFAULT_SCOPE));
        String error = Json.string(response, "error", "");
        if (!error.isEmpty()) {
            throw new AuthException("Microsoft: " + Json.string(response, "error_description", error));
        }
        Account refreshed = finishLogin(Json.string(response, "access_token", ""),
                Json.string(response, "refresh_token", account.msaRefreshToken), message -> { });
        refreshed.id = account.id;
        return refreshed;
    }

    private Account finishLogin(String msaAccessToken, String msaRefreshToken, Consumer<String> progress) {
        if (msaAccessToken.isBlank()) {
            throw new AuthException("Microsoft не выдал access token");
        }

        progress.accept("Xbox Live…");
        JsonObject xblProperties = new JsonObject();
        xblProperties.addProperty("AuthMethod", "RPS");
        xblProperties.addProperty("SiteName", "user.auth.xboxlive.com");
        xblProperties.addProperty("RpsTicket", "d=" + msaAccessToken);
        JsonObject xblRequest = new JsonObject();
        xblRequest.add("Properties", xblProperties);
        xblRequest.addProperty("RelyingParty", "http://auth.xboxlive.com");
        xblRequest.addProperty("TokenType", "JWT");
        JsonObject xbl = postJson(XBL_URL, xblRequest, null);
        String xblToken = Json.string(xbl, "Token", "");

        progress.accept("XSTS…");
        JsonArray tokens = new JsonArray();
        tokens.add(xblToken);
        JsonObject xstsProperties = new JsonObject();
        xstsProperties.addProperty("SandboxId", "RETAIL");
        xstsProperties.add("UserTokens", tokens);
        JsonObject xstsRequest = new JsonObject();
        xstsRequest.add("Properties", xstsProperties);
        xstsRequest.addProperty("RelyingParty", "rp://api.minecraftservices.com/");
        xstsRequest.addProperty("TokenType", "JWT");
        JsonObject xsts = postJson(XSTS_URL, xstsRequest, null);
        String xstsToken = Json.string(xsts, "Token", "");
        String userHash = userHash(xsts);

        progress.accept("Minecraft services…");
        JsonObject mcRequest = new JsonObject();
        mcRequest.addProperty("identityToken", "XBL3.0 x=" + userHash + ";" + xstsToken);
        JsonObject mc = postJson(MC_LOGIN_URL, mcRequest, null);
        String mcAccessToken = Json.string(mc, "access_token", "");
        long expiresIn = Json.number(mc, "expires_in", 86400);

        JsonObject entitlements = getJson(MC_ENTITLEMENTS_URL, mcAccessToken);
        if (entitlements.has("items") && entitlements.getAsJsonArray("items").isEmpty()) {
            throw new AuthException("На этом аккаунте нет купленного Minecraft: Java Edition");
        }

        progress.accept("Профиль…");
        JsonObject profile = getJson(MC_PROFILE_URL, mcAccessToken);
        String name = Json.string(profile, "name", "");
        String rawUuid = Json.string(profile, "id", "");
        if (name.isBlank() || rawUuid.isBlank()) {
            throw new AuthException("Аккаунт без профиля Minecraft — сначала создайте ник в лаунчере Mojang");
        }

        Account account = new Account();
        account.type = Account.Type.MICROSOFT;
        account.name = name;
        account.uuid = dashed(rawUuid);
        account.accessToken = mcAccessToken;
        account.accessExpiresAt = System.currentTimeMillis() + expiresIn * 1000L;
        account.msaRefreshToken = msaRefreshToken;
        if (profile.has("skins") && !profile.getAsJsonArray("skins").isEmpty()) {
            account.skinUrl = Json.string(profile.getAsJsonArray("skins").get(0).getAsJsonObject(), "url", null);
        }
        Log.info("Microsoft login complete for " + name);
        return account;
    }

    private static String userHash(JsonObject xsts) {
        if (xsts.has("DisplayClaims")) {
            JsonObject claims = xsts.getAsJsonObject("DisplayClaims");
            if (claims.has("xui") && !claims.getAsJsonArray("xui").isEmpty()) {
                return Json.string(claims.getAsJsonArray("xui").get(0).getAsJsonObject(), "uhs", "");
            }
        }
        throw new AuthException("Xbox Live не вернул user hash (возможно, детский аккаунт без разрешений)");
    }

    private static String dashed(String raw) {
        if (raw.contains("-")) {
            return raw;
        }
        return raw.replaceFirst("(\\p{XDigit}{8})(\\p{XDigit}{4})(\\p{XDigit}{4})(\\p{XDigit}{4})(\\p{XDigit}{12})",
                "$1-$2-$3-$4-$5");
    }

    private static void sleep(int seconds) {
        try {
            Thread.sleep(seconds * 1000L);
        } catch (InterruptedException interrupted) {
            Thread.currentThread().interrupt();
            throw new AuthException("Вход прерван");
        }
    }
}
